### PR TITLE
Consensus tx clean-up in tx pool & fix bugs

### PR DESCRIPTION
**Pull request by liayoo**
- Added consensus tx clean-up for discarded blocks
- Fixed stakeTx not setting bug
- Updated to return fast if block is already notarized in tryUpdateNotarized()
- Added a nonce check to isValidConsensusTx()
- Reduced MAX_CONSENSUS_LOGS_IN_STATES to 100 and renamed as MAX_INVALID_BLOCKS_ON_MEM
- Updated to try to get block from the on-mem chain by hash in getSnapDb().

### CODE REVIEW

File: block-pool/index.js [hunk 1]
```diff
@@ -468,6 +471,12 @@ class BlockPool {
   }
 
   cleanUpForBlockHash(blockHash) {
+    if (_get(this.hashToBlockInfo[blockHash], 'block')) {
```
Comment (platfowner) [inline block-pool/index.js:474]: minor: what about adding variable definition e.g. `const block = _get(this.hashToBlockInfo[blockHash], 'block');`?

File: block-pool/index.js [hunk 2]
```diff
@@ -468,6 +471,12 @@ class BlockPool {
...
+      this.node.tp.cleanUpConsensusTxsForBlock(this.hashToBlockInfo[blockHash].block);
+    }
+    if (_get(this.hashToInvalidBlockInfo[blockHash], 'block')) {
```
Comment (platfowner) [inline block-pool/index.js:477]: minor: ditto?

File: tx-pool/index.js [hunk 3]
```diff
@@ -410,6 +410,51 @@ class TransactionPool {
...
+    }
+  }
+
+  cleanUpConsensusTxsForBlock(block) {
+    const consensusTxs = new Set(block.last_votes.map((tx) => tx.hash));
+    this.addEvidenceTxsToTxHashSet(consensusTxs, block.evidence);
+    this.updateTxPoolWithTxHashSet(consensusTxs, {}, {});
+  }
+
   cleanUpForNewBlock(block) {
```
Comment (platfowner) [inline tx-pool/index.js:458]: could you update the test case in tx-pool.test.js by adding some evidences to blocks?

**Review by platfowner — APPROVED** [review]
LGTM with some comments.
Thanks!

File: tx-pool/index.js (hunk 3 above)
Comment (liayoo) [reply tx-pool/index.js:461]: Added a test case for cleanUpConsensusTxsForBlock() with some evidence votes. Thanks!

File: block-pool/index.js (hunk 1 above)
Comment (liayoo) [reply block-pool/index.js]: done, thanks!

File: block-pool/index.js (hunk 2 above)
Comment (liayoo) [reply block-pool/index.js]: ditto

**Comment by liayoo** [comment]
Thanks for the review~ Now merging!

**Pull request merged**